### DISCUSSION TITLE
Docker Samples - Nuget Caching - Why is this a separate build step in all the examples?

### DISCUSSION THREAD

**Discussion by bt-Knodel** [dotnet/dotnet-docker#6123] [Q&A]
I've reviewed the different samples and most samples with Docker suggest putting the nuget restore into a separate build stage, copying only the *.csproj files to perform the restore.

However, I'm curious why it's done this way.  I've found that simply declaring the global nuget caching folder as an explicit cache mount results in extremely similar build performance while simplifying the setup for multi-folder/multi-project solutions.

Is there a particular reason all the samples do that or is this perhaps another example to share with the community?  We found this particular setup worked well for our fairly large project (20ish project references with a custom secured nuget repository).  I did see this question was similar, but didn't offer the solution we ultimately landed on as an alternative approach: [Previously posted question](https://github.com/dotnet/dotnet-docker/discussions/4187#discussion-4540394)

**Dockerfile example caching nuget without separating from publish:**
```
FROM mcr.microsoft.com/dotnet/aspnet:8.0 AS base

# Runtime Dependencies

FROM mcr.microsoft.com/dotnet/sdk:8.0 AS build

WORKDIR /src
COPY . .

RUN --mount=type=cache,id=nuget,target=/root/.nuget/packages \ 
        --mount=type=cache,id=dotnet_tools,target=/root/.dotnet \
        dotnet publish "SampleApiProject.csproj"

FROM base AS runtime
WORKDIR /app
COPY --from=build /app/publish .

ENTRYPOINT ["dotnet", "SampleApiProject.dll"]
```

**Comment by akoken** [comment]
It’s a best practice because each `RUN` command in your Dockerfile creates its own layer, and `Docker` caches these layers. When you update your code but don’t touch the dependencies in your .csproj file, Docker skips rebuilding the dependency layer and uses the cached version instead. You can even see this in the output of the docker build command, where it shows which layers are reused.

**Reply by bt-Knodel** [reply]
I understand it skips the nuget layer.  But in more complex solutions, especially mono-repos, there's numerous startup projects, nuget.config's, directory.build.props, and nuget lock files.

The needed files for a nuget restore generally require far more than just the csproj, and this has a fantastic convenience associated to it since you don't have to track all those build assets in every unique startup project Dockerfile.

**Comment by lbussell** [comment]
Yes, using cache mounts for NuGet packages is a good solution and it can offer similar or better performance to using Docker's layer caching for local dev workflows. The benefit is amplified when you are building multiple projects that share dependencies.

You can even cache .NET's build outputs like so:
```
  --mount=type=cache,target=/src/bin \
  --mount=type=cache,target=/src/obj \
```

As for why it's not the default in the samples, I think it's a combination of:
- `RUN --mount=...` doesn't work on Windows yet since it requires BuildKit. And when this pattern was established, we definitely weren't yet using BuildKit on Linux.
- We simply haven't got around to it. There was an effort bundled into another PR a while back to add cache mounting to some of the samples: https://github.com/dotnet/dotnet-docker/pull/4951. It just never got merged.

We would happily take community contributions in this area (Issues or Pull Requests)!

**Reply by bt-Knodel** [reply]
Fantastic!  I was unaware.  I have been interested in helping contribute.  If I were to open an issue/PR, would it be better side-by-side with the existing example dockerfiles or updating existing dockerfiles?  I have some time next week to put something in if you're willing to provide a bit of guidance on where the samples belong!

Side note, is there a particular reason you mention the context "local dev workflows"?  We use this dockerfile setup in pipelines/deploys too and it works well.

**Reply by lbussell** [reply]
> is there a particular reason you mention the context "local dev workflows"?

If your build pipeline is just one project and one Dockerfile (as is the case for most of our samples), I think the cache mount provides limited benefit. If you get a clean build/CI machine when you build and deploy, then you'll be building the whole Dockerfile with an empty cache every time. Whereas, if you are developing your app locally and are repeatedly rebuilding the image, then you'll the build speed improvement. This is all assuming you haven't set up an [external cache](https://docs.docker.com/build/cache/optimize/#use-an-external-cache). Multi-project or multi-Dockerfile solutions will see the biggest benefit.

> If I were to open an issue/PR, would it be better side-by-side with the existing example dockerfiles or updating existing dockerfiles?

I think that we would want to update some or all of the existing Linux sample Dockerfiles. The [`complexapp`](https://github.com/dotnet/dotnet-docker/tree/main/samples/complexapp) sample may be a good place to start, since it has just one Dockerfile with multiple projects. With some of the other samples, like `aspnetapp` and `dotnetapp`, we'd want to make sure that the caching works correctly with [cross-platform builds](https://github.com/dotnet/dotnet-docker/blob/main/samples/build-for-a-platform.md#multi-platform-dockerfiles-with---platform). I don't think Self-contained or AOT publishing should be affected.
